### ISSUE TITLE
Path resolution issue

### ISSUE THREAD

**Issue by mkarajohn** [requirejs/requirejs#1607]
I am working on a project that uses `npm` for dependency management. I have declared all the dependencies of the project in the `paths` option of the config. In essence I have mapped my `package.json` dependencies to the `paths` object, 1 to 1, with each entry pointing to the main source file. So for example, my `paths` looks like this.

```
requirejs.config({
    baseUrl: '/static/',
    paths: {
        packageA: 'node_modules/packageA/index'
    }
});
```

and my `node_modules` structure is like this:

```
node_modules/
|- packageA/
    |- addons/
        |- someAddon.js
    |- index.js
```

This way I can simply define `packageA` as a dependency for other modules, instead of the more verbose `'node_modules/packageA/index'`. (assume my `node_modules` file is accessible from the browser, like all other static files)

This works.

## The problem

Let's say that I want to use one of the optional addons that come with `packageA`. In order to do that I must `require` or pass as a dependency in the `define` of another module the path pointing to the addon file. For example:

```
define(['packageA', 'node_modules/packageA/addons/someAddon'], function(packageA) { 
    /* Use the enhanced packageA */ 
});
```

Now, the way the addon will require the package's main file will most likely be something like this:

```
define(['../index'], function(packageA) { /* add functionality to package */ });
```

It will require the main file relative to the addon file, as is normal. 

However in the case of the example from above:

```
define(['packageA', 'node_modules/packageA/addons/someAddon'], function(packageA) {
    /* Use the enhanced packageA */ 
});
```

This will not work. The page will hang and will never load all my dependencies.

After spending literally dozens of hours trying to figure out why my app would not load, with no error messages whatsoever, (first thinking it was due to cyclic dependencies and later not knowing what to think) I came to realise that it's because of the fact that in my modules I am requiring the package by the `paths` defined handle, AKA simply by `packageA`, while the addons will require it by relative path. 

If in the example above I do

```
define(['node_modules/packageA/index', 'node_modules/packageA/addons/someAddon'], function(packageA) {
    /* Use the enhanced packageA */ 
});
``` 

then everything will work. 

## To sum up

For some reason requiring a module by its `paths` alias, in some modules, while also requiring by relative path in some other modules will break things silently.

**Comment by jrburke** [comment]
For npm packages, it is best to use either [packages config](http://requirejs.org/docs/api.html#packages) or use a tool like [adapt-pkg-main](https://github.com/jrburke/adapt-pkg-main) to write out some packagename modules that point to the "main" modules in the package.

Closing as a discussion ticket, but feel free to continue discussion here.

**Comment by jrburke** [comment]
Depending on how the modules in the package reference each other, you may also need to set the [nodeIdCompat config option](http://requirejs.org/docs/api.html#config-nodeIdCompat).

**Comment by mkarajohn** [comment]
Thanks for the tip. 

Simply registering the package as a `package` object resolved the issue. 

I noticed that when I `require`d the main package by its `package` name, and then I `require`d the addons by pointing to their absolute path, the main package would get loaded but the addons would not be applied to it. 

Just for the record, can you explain to me what was happening when I was trying to load my packages the way described in the op and when trying to load them the way I described in the paragraph above?

If the explanation is too complex and you don't feel like writing, I am fine with a simplified version.

**Comment by mkarajohn** [comment]
Also, now that I come to think of it I am not sure what the difference between 'paths` and `packages` is, except that `packages` is like a better `paths`?..

**Comment by jrburke** [comment]
When loading with npm packages it is best to not use absolute paths, but rather module IDs that are based on the package names, so `packageA/addons/someAddon` instead of `nodepackageA/addons/someAddon`, that should resolve better.

paths config is basically setting a path prefix based on a module ID prefix. Packages config is different because packages normally want a require for `packageName` to actually be the "main" module inside the package, but also allow other non-"main" modules to be loaded using the `packageName` as the start of the module ID segment. The different rule around the "main" ID resolution leads to packages config being different from paths config.

**Comment by ScottFreeCode** [comment]
If I understand correct, `packages` name and location are equivalent to `paths` entries in that if you set them from the package name to the package directory (rather than the package's main module) then they'll properly handle all the sub-modules and relative paths involving that package, but `packages` differs in also offering a mapping of the package name (as a complete module ID, not affecting sub-modules) to the package main module (much the way Node automatically resolves `index.js` or whichever entry point is in `package.json`, if I recall correct); is that accurate?

**Comment by chrisknoll** [comment]
I believe if you were to look at how modules were registered and you had your package 'mypackage' set up using the default main (so it will grab main.js by default), if you did this:

```
define(['mypackage'], function (myPackage) {} )

The main.js module under your package folder mypackage/main.js would be loaded, the moduleID given to it would be mypackage/main, and anything that main loaded up if it was using relative moduleIDs would be mypackage/submodule1, mypackage/submodule2 and submodule1.js and submodule2.js would be located in the same folder under your mypackage folder and siblings of the main.js.  I believe how that all works.

I think the main diff between pacakges and paths is packages resolves to a top-level moduleID, and all modules within the package are assumed to be under that root moduleID.  

In other words, defining a package 'mypackage' doesn't locate a module called 'mypackage', unlike a paths config if you said 'mypackage': '/path/to/mypackage' that would actually load up a js file called mypackage.js and associate a module id to that 'mypackage' id. 

From a r.js optimization perspective, If you wanted to bundle all the stuff under a package into a standalone omptimized js file, i think you'd just set up a r.js config to make the main module to build the package and it will just compile the whole thing into 1 file, but in this case, i am not sure how you reference the optimized package file...does it now become something in the paths config so that it loads the optimized .js file, or do you still refer to it as a package?  @jrburke, any comments on packages and how they'd be loaded after optimization?

**Comment by jrburke** [comment]
> any comments on packages and how they'd be loaded after optimization?

The r.js optimizer will inline a define() for the 'packagename', and it will have a dependency on the "main" module ID in the package. This allows for scenarios that do not have a package config set up for it after a build. For example, if the build bundles [almond](https://github.com/jrburke/almond) for the AMD API provider, it does not need the package config included in the built file.

**Comment by chrisknoll** [comment]
Ok, thanks for the info, I have a follow up question: if i use r.js optimizer to bundle up one package (which might have dozens of assents inside it so the optimization is a win there), how does the post-optimization require config change such that when i see a requirejs(['mypackage'], function(mypackage) {} (or if i want to create another AMD module that depends on 'mypackage' and i've written it like: define(['mypackage'], function(mypackage){}, does that require a package config to be set up, and if so, how does the optimized 'mypackage' get loaded up by the applcation that want's to reference the optimized build of 'mypackage'?

My guess is: i need to (for the dependent application that wants to incorporate a built-optimized 'mypackage' module, maybe hosted on a remote CDN) create a paths config that points 'mypackage/main' to the built 'mypackage.min.js' file, which will then load up all the other dependent submodules in there (which is inlined).  

The crux of my question is: if i have a package in my web app which I'd like to just optimize it individually into a stand-alone .js file, what sort of configuration changes to i need to make inside the web app so that it understand that doing require("mypackage") means it should load up the optimized bundle (and therefore register all the sub-modules that are included in the package)?  Taking that a step further, since you are able to reference a submodule in the package as 'mypackage/submodule', if I require("mypackage/submodule") directly, how will it know to go to 'mypacakge.min.js' to find it?

To try to answer my own question: in the built verson of the web applcation that wants to reference the optimized 'mypackage', do I change the location: property of the package config to point to the optimized .js?

Last point:  the idea here is to take the different packages of components and optimize them individually so that they can be used between different applications (think of this as a 'shared package').  Part of this sharing is to assert that any consumer of these packages will have to provide the AMD module loader (I wouldn't bundle each of these shared compoents with an AMD provider is what i'm saying).  I don't think this is a problem constraint for my user base: if they want to use these packages, they need to do it inside a AMD based container.

I understand that there's a require.js documentation on sharing a module between 2 pages. This is similar to that exercise, but this case uses **packages**, not optimized module bundles.  Not sure if this difference is significance, but I just would like to get my confusion cleared up about the implications of a require.js config for packages in the context of an optimized bundle.

**Comment by jrburke** [comment]
If the built file is called packagename.js, only contains modules from that package, and the goal is to use the main module for that package from that built file, then the consuming app only needs to be able to resolve 'packagename' to that packagename.js file. 

If the built file packagename.js is placed in the baseUrl for the project, then no config is needed. Otherwise paths config may be needed. However, it will not work to use packages config in that particular built case if packagename.js is dynamically loaded by the loader.
